### PR TITLE
[SPARK-31756][WEBUI] Add real headless browser support for UI test

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR mainly adds two things.

1. Real headless browser support for UI test
2. A test suite using headless Chrome as one instance of  those browsers.

Also, for environment where Chrome and Chrome driver is not installed, `ChromeUITest` tag is added to filter out the test suite.
By default, test suites with `ChromeUITest` is disabled.

### Why are the changes needed?

In the current master, there are two problems for UI test.
1. Lots of tests especially JavaScript related ones are done manually.
Appearance is better to be confirmed by our eyes but logic should be tested by test cases ideally.
 
2. Compared to the real web browsers, HtmlUnit doesn't seem to support JavaScript enough.
I added a JavaScript related test before for SPARK-31534 using HtmlUnit which is simple library based headless browser for test.
The test I added works somehow but some JavaScript related error is shown in unit-tests.log.

```
======= EXCEPTION START ========
Exception class=[net.sourceforge.htmlunit.corejs.javascript.JavaScriptException]
com.gargoylesoftware.htmlunit.ScriptException: Error: TOOLTIP: Option "sanitizeFn" provided type "window" but expected type "(null|function)". (http://192.168.1.209:60724/static/jquery-3.4.1.min.js#2)
        at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$HtmlUnitContextAction.run(JavaScriptEngine.java:904)
        at net.sourceforge.htmlunit.corejs.javascript.Context.call(Context.java:628)
        at net.sourceforge.htmlunit.corejs.javascript.ContextFactory.call(ContextFactory.java:515)
        at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.callFunction(JavaScriptEngine.java:835)
        at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.callFunction(JavaScriptEngine.java:807)
        at com.gargoylesoftware.htmlunit.InteractivePage.executeJavaScriptFunctionIfPossible(InteractivePage.java:216)
        at com.gargoylesoftware.htmlunit.javascript.background.JavaScriptFunctionJob.runJavaScript(JavaScriptFunctionJob.java:52)
        at com.gargoylesoftware.htmlunit.javascript.background.JavaScriptExecutionJob.run(JavaScriptExecutionJob.java:102)
        at com.gargoylesoftware.htmlunit.javascript.background.JavaScriptJobManagerImpl.runSingleJob(JavaScriptJobManagerImpl.java:426)
        at com.gargoylesoftware.htmlunit.javascript.background.DefaultJavaScriptExecutor.run(DefaultJavaScriptExecutor.java:157)
        at java.lang.Thread.run(Thread.java:748)
Caused by: net.sourceforge.htmlunit.corejs.javascript.JavaScriptException: Error: TOOLTIP: Option "sanitizeFn" provided type "window" but expected type "(null|function)". (http://192.168.1.209:60724/static/jquery-3.4.1.min.js#2)
        at net.sourceforge.htmlunit.corejs.javascript.Interpreter.interpretLoop(Interpreter.java:1009)
        at net.sourceforge.htmlunit.corejs.javascript.Interpreter.interpret(Interpreter.java:800)
        at net.sourceforge.htmlunit.corejs.javascript.InterpretedFunction.call(InterpretedFunction.java:105)
        at net.sourceforge.htmlunit.corejs.javascript.ContextFactory.doTopCall(ContextFactory.java:413)
        at com.gargoylesoftware.htmlunit.javascript.HtmlUnitContextFactory.doTopCall(HtmlUnitContextFactory.java:252)
        at net.sourceforge.htmlunit.corejs.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3264)
        at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$4.doRun(JavaScriptEngine.java:828)
        at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$HtmlUnitContextAction.run(JavaScriptEngine.java:889)
        ... 10 more
JavaScriptException value = Error: TOOLTIP: Option "sanitizeFn" provided type "window" but expected type "(null|function)".
== CALLING JAVASCRIPT ==
  function () {
      throw e;
  }
======= EXCEPTION END ========
```
I tried to upgrade HtmlUnit to 2.40.0 but what is worse, the test become not working even though it works on real browsers like Chrome, Safari and Firefox without error.
```
[info] UISeleniumSuite:
[info] - SPARK-31534: text for tooltip should be escaped *** FAILED *** (17 seconds, 745 milliseconds)
[info]   The code passed to eventually never returned normally. Attempted 2 times over 12.910785232 seconds. Last failure message: com.gargoylesoftware.htmlunit.ScriptException: ReferenceError: Assignment to undefined "regeneratorRuntime" in strict mode (http://192.168.1.209:62132/static/vis-timeline-graph2d.min.js#52(Function)#1)
```
To resolve those problems, it's better to support headless browser for UI test.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

I tested with following patterns. Both Chrome and Chrome driver should be installed to test.

1. sbt / with default excluded tags (ChromeUISeleniumSuite is expected to be skipped and SQLQueryTestSuite is expected to succeed)
`build/sbt -Dspark.test.webdriver.chrome.driver=/path/to/chromedriver "testOnly org.apache.spark.ui.ChromeUISeleniumSuite org.apache.spark.sql.SQLQueryTestSuite" 

2. sbt / overwrite default excluded tags as empty string (Both suites are expected to succeed)
`build/sbt -Dtest.default.exclude.tags= -Dspark.test.webdriver.chrome.driver=/path/to/chromedriver "testOnly org.apache.spark.ui.ChromeUISeleniumSuite org.apache.spark.sql.SQLQueryTestSuite" 

3. sbt / set `test.exclude.tags` to `org.apache.spark.tags.ExtendedSQLTest` (Both suites are expected to be skipped)
`build/sbt -Dtest.exclude.tags=org.apache.spark.tags.ExtendedSQLTest -Dspark.test.webdriver.chrome.driver=/path/to/chromedriver "testOnly org.apache.spark.ui.ChromeUISeleniumSuite org.apache.spark.sql.SQLQueryTestSuite" 

4. Maven / with default excluded tags (ChromeUISeleniumSuite is expected to be skipped and SQLQueryTestSuite is expected to succeed)
`build/mvn -Dspark.test.webdriver.chrome.driver=/path/to/chromedriver -Dtest=none -DwildcardSuites=org.apache.spark.ui.ChromeUISeleniumSuite,org.apache.spark.sql.SQLQueryTestSuite test`

5. Maven / overwrite default excluded tags as empty string (Both suites are expected to succeed)
`build/mvn -Dtest.default.exclude.tags= -Dspark.test.webdriver.chrome.driver=/path/to/chromedriver -Dtest=none -DwildcardSuites=org.apache.spark.ui.ChromeUISeleniumSuite,org.apache.spark.sql.SQLQueryTestSuite test`

6. Maven / set `test.exclude.tags` to `org.apache.spark.tags.ExtendedSQLTest` (Both suites are expected to be skipped)
`build/mvn -Dtest.exclude.tags=org.apache.spark.tags.ExtendedSQLTest  -Dspark.test.webdriver.chrome.driver=/path/to/chromedriver -Dtest=none -DwildcardSuites=org.apache.spark.ui.ChromeUISeleniumSuite,org.apache.spark.sql.SQLQueryTestSuite test`

### CODE REVIEW

**Comment by sarutak** [comment]
retest this please.

**Comment by srowen** [comment]
Jenkins retest this please

File: dev/run-tests.py [hunk 1]
```diff
@@ -33,6 +33,9 @@
 from sparktestsupport.toposort import toposort_flatten
 import sparktestsupport.modules as modules
 
+always_excluded_tags = [
```
Comment (dongjoon-hyun) [inline dev/run-tests.py:36]: `always_excluded_tags` might be misleading because this should be tested in the proper environments.

File: pom.xml [hunk 2]
```diff
@@ -204,7 +204,7 @@
     <leveldbjni.group>org.fusesource.leveldbjni</leveldbjni.group>
 
     <test.java.home>${java.home}</test.java.home>
-    <test.exclude.tags></test.exclude.tags>
+    <test.exclude.tags>org.apache.spark.tags.ChromeUITest</test.exclude.tags>
```
Comment (dongjoon-hyun) [inline pom.xml:207]: Could you add some comment before this line?

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -1001,9 +1004,9 @@ object TestSettings {
     javaOptions += "-Xmx3g",
     // Exclude tags defined in a system property
     testOptions in Test += Tests.Argument(TestFrameworks.ScalaTest,
-      sys.props.get("test.exclude.tags").map { tags =>
-        tags.split(",").flatMap { tag => Seq("-l", tag) }.toSeq
-      }.getOrElse(Nil): _*),
+      sys.props.get("test.exclude.tags").map(tag => tag.split(",").toSeq)
+        .getOrElse(defaultExcludedTagsForScalaTest).filter(!_.trim.isEmpty)
```
Comment (dongjoon-hyun) [inline project/SparkBuild.scala:1008]: This looks not robust enough.
If a developer set an irrelevant tag (`ExtendedSQLTest` or `ExtendedHiveTest`), `ChromeUITest` suddenly will cause a failure in the irrelevant `Core` module.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @sarutak .
I left a few comments. Especially, could you take a look at https://github.com/apache/spark/pull/28627#discussion_r429671303 ?

As a downstream user, I'm using the other tags `ExtendedHiveTest` or `ExtendedSQLTest` a lot in my test environment. This had better be orthogonal to the other tags.

File: project/SparkBuild.scala (hunk 3 above)
Comment (sarutak) [reply project/SparkBuild.scala:1008]: It's what I intended that if developers set `test.excluded.tags` explicitly, all the tags which are intended to be excluded should be added explicitly.
But as you concerned, this change seems to break the build compatibility.
I'll submit another idea.

**Comment by sarutak** [comment]
Test patterns are updated in the description.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
Can those tests finish within 400m with Maven?

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
@dongjoon-hyun Is it possible to finish tests within 400m with Maven?
I see #28640 also fails tests with Maven because of time out.
Can we extend the time for tests with Maven somehow?

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
@HyukjinKwon @shaneknapp 
The time limit for CI seems to be too short with Maven.
Tests for this PR has failed 9 times due to the limit.
I see [another PR](https://github.com/apache/spark/pull/28640#issuecomment-634304376) fails with the same reason. 
Can we expand the limit for Maven or do you have better idea?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
The problem is that it's getting too long to finish each PR build .. so last time it was pretty arguable when it increased the build time IIRC. Maybe it's time to profile and reduce the testing time again .. adding @cloud-fan too FYI. Shall we increase a bit more for now since it blocks Maven testing as can be seen above?

**Comment by sarutak** [comment]
Couldn't finish within 500m...

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
It seems that we are hitting time-out still in some cases.

BTW, @sarutak . Is the following a successful run?
- https://github.com/apache/spark/pull/28627#issuecomment-635776114

**Comment by sarutak** [comment]
> BTW, @sarutak . Is the following a successful run?

Actually, It's not. 
According to the console log, it's almost finished but aborted...
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123247/console

**Comment by dongjoon-hyun** [comment]
Oh. Got it, @sarutak .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for updating the PR by introducing `test.default.exclude.tags`.
Merged to master.

**Comment by sarutak** [comment]
Thanks all for the review and expanding the timeout!

**Comment by shaneknapp** [comment]
sorry, i don't watch github for @mentions...  if this is still a problem,
file a build/infra jira and assign it to me.  :)

**Pull request closed without merging**